lib/logstash/inputs/unix.rb

```diff
@@ -118,6 +117,35 @@ def handle_socket(socket, output_queue)
     end
   end
 
+  ##
+  # Emulates `IO#readpartial` with a timeout and our plugin's stop-condition,
+  # limiting blocking calls to windows of 10s or less to ensure it can be interrupted.
+  #
+  # @param readable_io [IO] the IO to read from
+  # @param maxlen [Integer] the max bytes to be read
+  # @param timeout [Number] the maximum number of seconds to , or -1 to disable timeouts
+  #
+  # @return [:data_timeout] if timeout was reached before bytes were available
+  # @return [:stopping] if plugin stop-condition was detected before bytes were available
+  # @return [String] a non-empty string if bytes became available before the timeout was reached
+  def io_interruptable_readpartial(readable_io, maxlen, timeout)
+
+    data_timeout_deadline = timeout < 0 ? nil : Time.now + timeout
+    maximum_blocking_seconds = timeout < 0 || timeout > 10 ? 10 : timeout
+
+    loop do
+      return :stopping if stop?
+      result = readable_io.read_nonblock(maxlen, exception: false)
+
+      return result if result.kind_of?(String)
+      raise EOFError if result.nil?
```

Comment:
What would be a case where `esult.nil?` is true?

Reply:
Based on ]`IO#read_nonblock`](https://apidock.com/ruby/IO/read_nonblock) doc: 
At EOF, it will return nil instead of raising [EOFError](https://apidock.com/ruby/EOFError).

Reply:
Yes. When using `IO#read_nonblock` with `exception: false` we must handle all of the failure conditions ourselves.